### PR TITLE
[mlir][transform] Fix new interpreter and library preloading passes.

### PR DESCRIPTION
**Pull request by ingomueller-net**
This PR fixes the two recently added passes from #68661, which were non-functional and untested. In particular:
* The passes did not declare their dependent dialects, so they could not run at all in the most simple cases.
* The mechanism of loading the library module in the initialization of the intepreter pass is broken by design (but, fortunately, also not necessary). This is because the initialization of all passes happens before the execution of any other pass, so the "preload library" pass has not run yet at the time the interpreter pass gets initialized. Instead, the library is now loaded every time the interpreter pass is run. This should not be exceedingly expensive, since it only consists of looking up the library in the dialect. Also, this removes the library module from the pass state, making it possible in the future to preload libraries in several passes.
* The PR adds tests for the two passes, which were completely untested previously.

### CODE REVIEW

File: mlir/include/mlir/Dialect/Transform/Transforms/Passes.td [hunk 1]
```diff
@@ -51,8 +51,9 @@ def PreloadLibraryPass : Pass<"transform-preload-library"> {
 
     Warning: Only a single such pass should exist for a given MLIR context.
     This is a temporary solution until a resource-based solution is available.
-    TODO: investigate using a resource blob if some ownership mode allows it.
   }];
+  // TODO: investigate using a resource blob if some ownership mode allows it.
+  let dependentDialects = ["transform::TransformDialect"];
```
Comment (ftynse) [inline mlir/include/mlir/Dialect/Transform/Transforms/Passes.td:56]: Nit: use fully-qualified names in .td, e.g., `::mlir::transform::TransformDialect`.

File: mlir/test/Dialect/Transform/preload-library.mlir [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+
+// Note: no remark here since local entry point takes precedence.
+module attributes { transform.with_named_sequence } {
+  transform.named_sequence @private_helper(!transform.any_op {transform.readonly}) {
+  ^bb0(%arg0: !transform.any_op):
+    // expected-remark @below {{applying transformation}}
+    transform.test_transform_op
+    transform.yield
+  }
+}
```
Comment (ftynse) [inline mlir/test/Dialect/Transform/preload-library.mlir:19]: Nit: training newline.

**Review by ftynse — COMMENTED** [review]
Could you elaborate why the dialect dependency is necessary for these passes? It is usually required when the pass creates entities from the dialect, are these passes creating something?

**Comment by joker-eph** [comment]
> The PR adds tests for the two passes, which were completely untested previously.

Independently of this PR, I'm concerned by this, I'll comment on the original PR!

**Comment by ingomueller-net** [comment]
Thanks for the review. Nits addressed in latest commit.

> Could you elaborate why the dialect dependency is necessary for these passes? It is usually required when the pass creates entities from the dialect, are these passes creating something?

Sure:
* The preload pass calls [`MLIRContext::getOrLoadDialect<transform::TransformDialect>`](
https://github.com/llvm/llvm-project/blob/1bf08709/mlir/lib/Dialect/Transform/Transforms/PreloadLibraryPass.cpp#L35C22-L35C69).
* The interpreter pass [calls `transform::detail::getPreloadedTransformModule`](https://github.com/llvm/llvm-project/blob/4606712ef5b422edbe3799b665dcad7dcf348b90/mlir/lib/Dialect/Transform/Transforms/InterpreterPass.cpp#L30), which [does the same](https://github.com/llvm/llvm-project/blob/1bf08709/mlir/lib/Dialect/Transform/Transforms/TransformInterpreterUtils.cpp#L115C16-L115C63).

**Comment by joker-eph** [comment]
> Thanks for the review. Nits addressed in latest commit.

I don't think it was approved though, was it?

**Comment by ftynse** [comment]
> Thanks for the review. Nits addressed in latest commit.
> 
> > Could you elaborate why the dialect dependency is necessary for these passes? It is usually required when the pass creates entities from the dialect, are these passes creating something?
> 
> Sure:
> 
> * The preload pass calls [`MLIRContext::getOrLoadDialect<transform::TransformDialect>`](https://github.com/llvm/llvm-project/blob/1bf08709/mlir/lib/Dialect/Transform/Transforms/PreloadLibraryPass.cpp#L35C22-L35C69).
> * The interpreter pass [calls `transform::detail::getPreloadedTransformModule`](https://github.com/llvm/llvm-project/blob/4606712ef5b422edbe3799b665dcad7dcf348b90/mlir/lib/Dialect/Transform/Transforms/InterpreterPass.cpp#L30), which [does the same](https://github.com/llvm/llvm-project/blob/1bf08709/mlir/lib/Dialect/Transform/Transforms/TransformInterpreterUtils.cpp#L115C16-L115C63).

I see, thanks. What really escaped my attention here is that a pass cannot load a dialect because it run in multithreaded mode, so it has to declare as dependent any dialect it needed to load and have it loaded by the pass manager. This is different from creating entities from the dialect, which is impossible without the dialect being loaded in the first place. Not sure if this is documented sufficiently, somehow it wasn't obvious to me.

> I don't think it was approved though, was it?

Indeed, it wasn't. I haven't requested changes as I was rather asking for clarification. Approved after-the-fact, but generally better to wait for approval in the future.

**Comment by ingomueller-net** [comment]
> I see, thanks. What really escaped my attention here is that a pass cannot load a dialect because it run in multithreaded mode, so it has to declare as dependent any dialect it needed to load and have it loaded by the pass manager. This is different from creating entities from the dialect, which is impossible without the dialect being loaded in the first place. Not sure if this is documented sufficiently, somehow it wasn't obvious to me.

Thanks for the explanation! Those subtleties had escape *my* attention completely, so it's great to have them brought to my attention.

> > I don't think it was approved though, was it?
> 
> Indeed, it wasn't. I haven't requested changes as I was rather asking for clarification. Approved after-the-fact, but generally better to wait for approval in the future.

So sorry!! This wasn't on purpose -- I had assumed/thought that it was approved, which it wasn't. Will be more careful next time!

**Pull request merged**